Update overload docs

Michael0x2a opened the pull request:
This pull request:

1.  Moves documentation about overloads into the "More types" section.

    (Rationale: although overloads aren't really a type, I feel it makes sense for them to live somewhere in the "Type System Reference" section instead of the "Miscellaneous" section.)

2.  Modifies the docs to start with a non-dunder example (resolves https://github.com/python/mypy/issues/4579) and simplifies the `__dunder__` example.

3.  Adds some documentation about the "pick the first match" rule.

4.  Removes the now out-of-date note about erasure.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+This means that an overloaded function is still an ordinary Python
+function! There is no automatic dispatch handling: you must manually
+handle the different types in the implementation (usually by using
+``if`` statements and ``isinstance`` checks).
```
I would replace "usually" with "e.g."

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+ignored: they're overridden by the final implementation function.
+
+This means that an overloaded function is still an ordinary Python
+function! There is no automatic dispatch handling: you must manually
```
An exclamation sign is not needed here.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+function should be omitted: stubs do not contain runtime logic.
+
+Type checking calls to overloads
+--------------------------------
```
I would place the section about checking definitions first. I think it is more logical, one first writes a function, then one calls it.

ilevkivskyi replied:
> And the examples above already hint at how the definitions are checked.

This way the logic actually jumps IMO. We start from definitions in introduction, then we switch to checking calls, then then back to definitions. But this is not something critical, I can live with current ordering :-)

gvanrossum replied:
Such jumping logic is often more readable than an exhaustive treatment per sub-topic -- this is probably different from how you've learned to write mathematical proofs. :-)

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+
+If multiple variants end up matching a call, mypy will, for the
+most part, select the return type corresponding to the first
+matching call. For example, consider the following program:
```
I would reformulate this slightly. We can say something like: "Mypy will chose the best matching variant (considering argument kinds and types). If there are more than one equally good match, mypy will choose the first one. An exception is when there are multiple matching due to an `Any` type of some of the arguments, in this case mypy will infer `Any` as the result of the call."

ilevkivskyi replied:
> Is there still a gradient in how "good" a match is?

Yes, we still use some heuristics, we discussed for some time and agreed it would be impossible to specify all corner cases in a way making everyone happy. So we deliberately leave some room for type checkers in ambiguous corner cases (the cases are quite well specified, but the behavior in each is up to type checker).

Also, @Michael0x2a how about you prepare a PR to peps repo using your original proposal on typing tracking and taking into account the things you learned from implementing all this?

> And what about provided context? Does that play a role?

Currently it is used only if there type variables I think. I think in future we can use it to pick up s better overload, but it is more complex that it seems.

gvanrossum replied:
> the cases are quite well specified

Remind me where? PEP 484 explains almost entirely through examples. And the mypy docs don't apply to other type checkers. Are the heuristics only relevant to Protocols (PEP 544)? Then that should be called out specifically.

ilevkivskyi replied:
> Remind me where?

Original PEP 484 amendment proposal https://github.com/python/typing/issues/253#issuecomment-389262904

Some amendments to amendment discussion is also here https://github.com/python/mypy/issues/5119

> Are the heuristics only relevant to Protocols (PEP 544)?

No, there are several other corner cases: empty collections, star args, union types.

ilevkivskyi replied:
Also beware, the original proposal is _very_ long. It also discusses motivation, but the PEP 484 PR should include mostly the specification part IMO.

gvanrossum replied:
Possibly it can become a separate PEP, and PEP 484 should be amended with a link to the new PEP? That shouldn't be a big deal (it's not like type checking PEPs seem to attract a lot of attention from the usual crowd of armchair critics).

ilevkivskyi replied:
> Possibly it can become a separate PEP, and PEP 484 should be amended with a link to the new PEP?

I have no preferences, I original thought it would be OK to add this to the PEP 484 because it already contains a section on overloads, but having another mini-PEP is also OK.

gvanrossum replied:
I expect evolving and reviewing the doc as a separate PEP will be easier. Otherwise it'll be a *very* large expansion to an already-approved PEP, and that doesn't sound completely fair to people who reviewed it originally.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+There are a few exceptions to the "pick the first match" rule.
+For example, if multiple variants match due to an argument
+being of type ``Any``, mypy will make the inferred type also
+be ``Any``.
```
I would replace this paragraph with an example showing a call with `Any` to an above function (especially if you reword as I suggest above).

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+
+Mypy will also prohibit you from writing overload variants that are
+inherently unsafely overlapping: for example, writing two variants
+that accept the same arguments but return different types.
```
I think this paragraph belongs to the below section.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+that accept the same arguments but return different types.
+
+Type checking the implementation
+--------------------------------
```
I would rename this to "Type checking the definition" and include more info here. In particular I would show two examples: one showing the "unsafely overlaps error" (e.g. with `int -> int` and `object -> str`), and one with "will be never matched" (briefly). For first example it would be great to explain the motivation with and example `x: object = 42` -> `f(x)` -> `BOOM!`

Michael0x2a replied:
Regarding this section/the surrounding text: I wonder if it might be better to just link to PEP 484, once the overload stuff is accepted there.

I'm currently trying to write the `int -> int` and `object -> str` example -- and it gets mired in the weeds pretty quickly. In particular, I'm finding it challenging to concisely explain why mypy flags this example as being unsafe yet accepts the `List[int] -> int` and `List[str] -> str` example despite the fact that we can trigger the exact same kind of error in both cases.

(More specifically, it'd be trivial for me to write an explanation that describe the design decisions behind this behavior in detail; I'm finding it non-trivial to write an explanation that's satisfying yet concise.)

ilevkivskyi replied:
> I'm finding it non-trivial to write an explanation that's satisfying yet concise.

Yes, I know this is the hardest part.

gvanrossum replied:
> I wonder if it might be better to just link to PEP 484, once the overload stuff is accepted there.

I'd rather be independent from that -- the purpose here is to be the best/clearest user documentation possible, while the goal of the PEP is to be as precise a specification as possible. Those are often conflicting goals.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
 
         name_by_id(3)  # int is not the same as UserId
 
+.. _function-overloading:
```
I would add a link to this section at the end of the `Callable` section.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+
+   To minimize potential issues, we recommend ordering your variants
+   from most to least specific. Your implementation should also
+   perform ``isinstance`` checks and the like in the same order
```
I would replace "``isinstance`` checks" to "runtime dispatch checks" (or something similar).

gvanrossum commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+``MyList`` example up above, the code in the body is checked with
+``index: Union[int, slice]`` and a return type ``Union[T, Sequence[T]]``.
+If there are no annotations on the implementation, then the body is
+not type checked.
```
Maybe clarify that this is different from having all args and return annotated with `Any`. That's standard, but somehow a reminder in this context seems helpful.

gvanrossum commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+.. note::
+
+   Due to the "pick the first match" rule, changing the order of your
+   overload variants can change how mypy type checks your program.
```
Maybe also mention that older versions of mypy (up to 0.610) used a different rule here? Or that PEP 484 is currently not clear on this?

gvanrossum commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+   overload variants can change how mypy type checks your program.
+
+   To minimize potential issues, we recommend ordering your variants
+   from most to least specific. Your implementation should also
```
How should I order them if they are all disjoint? (E.g. the ip_address example.)

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+    # type hints.
+    #
+    # Mypy will also check and make sure the signature is
+    # consistent with the provided variants.
```
> Did I do something wrong?

It looks like a bug. @Michael0x2a could you take a look, seems something simple.

Michael0x2a commented:
@ilevkivskyi and @gvanrossum -- I think this is ready for a second look whenever you have time.

Some notes:

1. I edited the overview at the top to include the text as requested
2. I tried switching the ip address example to something that wouldn't require us to talk about positional-only args, but wasn't really able to find anything I was happy with. (E.g. I thought about making an example where we return date vs datetime objects, but it ended up looking messier). I guess on the plus side, talking about positional-only args do hint to the read that types aren't the only thing that matters.
3. I switched to having the variant bodies use `...`, but decided against removing the blank lines between each variant: IMO it made the overloads look a little intimidatingly dense. That said, I don't feel too strongly about this issue, so I'm happy to remove them if people disagree.
4. I ended up making a new section about how overload variants are type checked. I couldn't really find a way to make it concise, unfortunately -- sorry about that.
5. Regarding the "we pick the best match" thing: I didn't really elaborate on it because I think the current behavior is mostly intuitive (?). If people disagree, I can spell out what exactly we mean by "best". I think it mostly boils down to talking about how star-args are handled?
6. I decided to keep the current topic order: I like using the [spiral approach](https://en.wikipedia.org/wiki/Spiral_approach) when teaching, and the topics basically do that: we start with examples and a high-level overview that communicate the gist of how to use overloads, then drill down into the nitty gritty semantics of calls, variant checks, and implementation checks (in that order).

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+Type checking the variants
+--------------------------
+
+Although the "pick the first match" algorithm works well in many cases, it can
```
This introductory sentence can make a confusing impression that this is some kind of limitation of mypys implementation. You can just say something simpler like "Not all possible variant combinations are considered safe by mypy. Let us consider an example of unsafe variants:"

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+        if isinstance(x, int):
+            return x
+        else:
+            return "Unsafe!"
```
It is not clear for a typical reader why it is unsafe at this point, use a different word here, maybe "Hm, not really a number".

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+
+If we examine just the annotated types, it seems as if ``bad_var`` ought to be of
+type ``str``. But if we examine the runtime behavior, ``bad_var`` will actually be
+of type ``int``!
```
Instead of saying the reasoning in words, it would be better to just show a code snippet, that will fail with a `TypeError` at runtime. For example, I would replace your snippet and subsequent explanation by this snippet:
```python
yolo: object = 42
unsafe_func(yolo) + "why not?"  # inferred type for the call is "str", but this will
                                # fail at runtime because the function will actually return 42

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+So in this example, the ``int`` argument in the first variant is a subtype of
+the ``object`` argument in the second, yet the ``int`` return type is a subtype of
+``str``. Both conditions are true, so mypy will correctly flag ``unsafe_func`` as
+being unsafe.
```
I really like these three paragraphs they nicely put on formal grounds the above example.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+``str``. Both conditions are true, so mypy will correctly flag ``unsafe_func`` as
+being unsafe.
+
+However, it is unfortunately impossible to detect *all* unsafe overloads without
```
All this long part about multiple inheritance is unnecessary I think. It can be replaced with a single short note at the end of this section like: "Note that mypy intentionally allows unsafe overloads related to multiple inheritance and empty containers. This way mypy avoids lots of false positives (that would otherwise make overloads impractical) by having some rare false negatives."

This is really a discussion that belongs to the PEP, not the docs. Remember that you write this mostly for people who have never heard about overloads, while currently it looks like you are talking to a CS prof: "Please forgive us for being so unsafe!" :-)

Michael0x2a replied:
Mm, I'm ok with just leaving a short snippet like that if we're ok with also linking to the PEP and telling the reader to read that if they want more details.

But I was under the impression that we wanted the docs to be more self-contained? In that case, I think we really do need at least one example of what an unsafe yet allowed overload could look like. I don't think a reader would really understand what we mean if with just that explanation .

One option I could try is re-using the `summarize` example instead -- that would let me cut out all of the setup related to multiple-inheritance which I think would help cut down the text by a decent amount. I still want to keep the gist of recommendations at the end and some of the exposition though. Empty collections are relatively common and some people do like to use mixins -- I think it's worth giving people a more detailed heads up.

ilevkivskyi replied:
> But I was under the impression that we wanted the docs to be more self-contained?

Self-contained doesn't mean complete, in particular here too much attention is devoted to motivation why we allow this. You could try re-using `summarize` if this will help cutting the size of this significantly.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+
+   Older versions of mypy used different semantics. In particular, overload
+   variants had to different after erasing types and calls to overloaded
+   functions did not use the "pick the first match" rule.
```
You can mention that mypy previously often silently returned `Any` for many overload calls. So that people will be not surprised by new errors when switching to 0.620.

Michael0x2a replied:
Will do

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+``Combined()`` is an instance of both ``A`` and ``B``; the empty list 
+``[]`` is an instance of both ``List[int]`` and ``List[str]``.
+
+
```
It looks like an example of "...will be never matched" is missing in this section.

Michael0x2a commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+    def ip_address(__a: int, __b: int, __c: int, __d: int) -> IPv4Address: ...
+    @overload
+    def ip_address(__a: int, __b: int, __c: int, __d: int,
+                   __e: int, __f: int, __g: int, __h: int) -> IPv6Address: ...
```
I think going with `ip_address(*args, **kwargs)` would ruin the narrative that overloads let you "refine" the type of your function to something more precise. Both options are also a kludge -- in that case, I'd rather use the kludge that teaches the reader something new and useful about mypy rather then handwaving the complication away.

That said, I think a better option would be to find a different example that avoids this issue completely. I'm just that I'm having difficulty coming up with a replacement that isn't totally contrived.

ilevkivskyi replied:
I just looked at mypy's own code and there is a single example of overload. It looks like this:
```python
@overload
def preferred_currency(c: PaidCustomer) -> Currency: ...
@overload
def preferred_currency(c: Customer) -> Optional[Currency]: ...
```

ilevkivskyi replied:
Implementation may be like this:
```python
def preferred_currency(c: Customer) -> Optional[Currency]:
    if isinstance(c, PaidCustomer):
        # Use currency from subscription plan
    elif in_app_purchases(c):
        # Use most commonly used currency
    else:
        return None
```

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+
+   To minimize potential issues, we recommend ordering your variants
+   from most to least specific. Your implementation should also
+   perform runtime checks (e.g. ``isinstance`` checks) in the same
```
Hm, now it seems to me it should be vice-versa here. People typically add annotations to existing code, so it is better to say "Your variants should be listed in the same order as runtime checks (e.g. `isinstance` checks) performed in the implementation."

gvanrossum commented:
I lost track -- who is waiting for whom here?

Michael0x2a commented:
@gvanrossum -- this PR is currently waiting on me. I've implemented most of Ivan's feedback but haven't had the chance to sit down and do the last bit of polishing yet.

Michael0x2a commented:
@ilevkivskyi and @gvanrossum -- Sorry for the delay! I think this should be ready for another look now.

Specific substantial changes made:

1. Swapped out the IP address example for something involving "mouse events". 

    (It's still a little contrived, but whatever -- I can showcase overloads w/ different arities, give an example of a call that matches the impl but not the variants, so good enough.)
2. Added an example/explanation of the "X will never be matched" error.
3. Replaced the example with multiple inheritance with an example about 'summarized' and tried making the exposition at the end less wordy/less "doom-and-gloom".
4. Fleshed out the section about old mypy semantics a little
5. Adjusted the section about recommendations for ordering overload variants to place more emphasis on making sure the runtime impl and variant orders mirror each other.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+        elif x2 is not None and y2 is not None:
+            return DragEvent(x1, y1, x2, y2)
+        else:
+            raise Exception("Bad arguments")
```
I think either `ValueError` or `TypError` will be better here.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+        elif x2 is not None and y2 is not None:
+            return DragEvent(x1, y1, x2, y2)
+        else:
+            raise Exception("Bad arguments")
```
Same here.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+.. note::
+
+   While we can leave the variant body empty using the ``pass`` keyword,
+   the more common convention is to instead use the ellipse (``...``) literal.
```
"ellipsis"

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+    def summarize(data: List[int]) -> int: ...
+
+    @overload
+    def summarize(data: List[str]) -> str: ...
```
How about having a signature that can't be expressed without overloads? For example, the first overload may be `List[int] -> float` and the implementation could return an average for integers, and a most common for strings.

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+    unsafe_func(some_obj) + " danger danger"  # Type checks, yet crashes at runtime!
+
+This program type checks according to the annotations, but will actually crash
+at runtime! Since ``some_obj`` is of type ``object``, mypy will decide that
```
The sentence "This program..." just repeats the comment above. I would just remove it. (Especially that we have an explanation in the next sentence.)

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+
+We run into a similar issue here. This program type checks if we look just at the
+annotations on the overloads. But since ``summarize(...)`` is designed to be biased
+towards returning an int when it receives an empty list, this program will actually
```
I would format this as ` ``int`` ` (or rather ` ``float`` ` if you update the signature as suggested).

ilevkivskyi commented on docs/source/more_types.rst:
```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+    def summarize(data: List[str]) -> str: ...
+
+    def summarize(data):
+        if len(data) == 0:
```
`if not data:`
